### ISSUE TITLE
Audio is super glitchy after installing GPIO touch screen.

### ISSUE THREAD

**Issue by tonketonky** [pideck/pideck#70]
Hi guys! 
I'm trying to set up PiDeck on my Raspberry Pi 3 using Audio Injector as a sound card. Everything worked smoothly when I used HDMI screen but once I installed driver for GPIO touch screen audio became very glitchy and also motion of the waveform isn't as fluent as when using HDMI screen. I'm using quite cheap touch screen I bought on ebay. I got the driver from their website here [http://www.kedei.net/raspberry/raspberry.html](url), it does something with the kernel but I don't see into these kernel things too much so I don't know, maybe it's just incompatible or something. Any ideas on this? Thanks in advance!

**Comment by obbardc** [comment]
It's probably the extra processing the touchscreen data requires, which
takes valuable processing power away from the sound card.

The HDMI graphics is handled separately by another part of the processor.

Which driver did you download, I will have another look into it...

We were thinking about getting a super dirt cheap one from eBay, maybe this
should go higher up the priority list to try out (Daniel?)



Cheers!

**Comment by tonketonky** [comment]
I downloaded LCD_show_v6_1_3 driver. 

So does it mean I'll probably have the same issue with a different GPIO touch screen as well? Do you suggest getting a HDMI one instead? 

Thanks!

And by the way big ups for this project! It's cool.

**Comment by obbardc** [comment]
Probably any others that go through the GPIO interface will be slow. HDMI
won't be an issue, and the official Raspberry Pi touchscreen works well,
and won't interact with your soundcard.

Let us know how you get on!

Cheers

**Comment by danielhjames** [comment]
Thanks Anton :-) The official Raspberry Pi touch screen uses a ribbon cable direct from the board instead of the GPIO pins. My best guess is that your touchscreen is trying to use the same interrupts as the audioinjector card.

For practical purposes you may find a QWERTY keyboard useful with the PiDeck, for example for text searches on your audio files. This means you don't really need a touchscreen, as all the xwax features are available by keyboard shortcuts. So any old PC monitor with the right cable would potentially do the job, as long as you don't need something really small and portable.

**Comment by tonketonky** [comment]
Yeah I know, I didn't express myself well, I was actually trying it with a keyboard and a TV via HDMI but I'd like to have PiDeck as part of my setup in rehearsal room where there already is a bunch of cables so yeah :D I want it small and a touch screen hat seemed to be a good choice to me to minimize the size.  

Anyways thanks a lot for your suggestions! This saved me from some further unsuccessful attempts with other screens hats. However I find official touchscreen quite larger than I need for this. I did some quick search now and I found 5" screens such as this one [https://www.elecrow.com/wiki/index.php?title=HDMI_Interface_5_Inch_800x480_TFT_Display](url) which connects to HDMI port for graphics and to GPIO for getting power and sending touch input. This shouldn't be an issue, right? Or should I completely avoid using GPIO when I use Audio Injector? 

Thanks

**Comment by danielhjames** [comment]
I would have thought sending just touch data over GPIO would not be too demanding, but we haven't tested that. I would also suggest reporting your initial findings to @flatmax the creator of the audioinjector, just in case it's useful to know for other audioinjector users.

**Comment by flatmax** [comment]
Interesting finding, glitchy audio. It is glitchy due to audio frames dropping, or distorted due to noise interference ?
Can you think of any reason why using an LCD with touch would create this problem ?
The only GPIO pins used by the original stereo audio injector card (and the zero) are I2C, power and I2S.

**Comment by obbardc** [comment]
I’d have thought it would be taking up a lot of cpu power to send the
display data, probably over spi or something

**Comment by flatmax** [comment]
If the Display and audio renders were on different threads, it would be 
possible to set priorities differently to favor scheduling of the audio 
over the video.


Matt

**Comment by tonketonky** [comment]
I meant glitchy in a meaning of frames dropping but after all it doesn't look like an issue just with outputting audio from the sound card itself but also with xwax interpreting the timecode input or something like that. 

First I tried just playing a wav file using aplay and the audio was perfectly clear. Then I removed all moving elements from xwax graphical interface and the audio became much better than when I was trying it yesterday but still not good enough. However now it doesn't sound like frames dropping but more like as if the vinyl spinning speed was fluctuating (which physically isn't, I tried it without touch screen and it worked great). Also it's significantly worse when playing A side than when playing B side. I'm using Serato Performance vinyl.  

And regarding using of GPIO pins the only two pins shared between audioinjector and touchscreen are 2 and 4 for power.

**Comment by danielhjames** [comment]
Could it be the audioinjector is starved of power by the touchscreen?

**Comment by obbardc** [comment]
On 4 January 2018 at 15:23, Daniel James <notifications@github.com> wrote:
>
> Could it be the audioinjector is starved of power by the touchscreen?

Doubtful, the audioinjector has a regulator on-board and is only
minimal current draw.


I think xwax uses a hell of a lot of CPU time, and that is interfering
with the SPI pins from the touch screen.

Can you disable the touch input in the kernel (not sure how to do
this) to isolate the touch/display communications?

Is there any way to set the display realtime priority to be much lower?



Cheers!

**Comment by tonketonky** [comment]
I managed to disable touch input by changing driver for touchscreen from "evdev" to "libinput" in /usr/share/X11/xorg.conf.d/10-evdev.conf. It doesn't make much of difference (if any) but I don't know if that's what you meant.

And why would you want to disable touch input? I thought that's not what makes an overload.

**Comment by tonketonky** [comment]
I made one more adjustment, turned volume on turntable to max, I probably had it a bit too low even though without touchscreen it worked at that level. Anyways now audio seems to be clear and with proper playing speed. Moving parts of xwax gui still need to be disabled but it's usable now!

**Comment by danielhjames** [comment]
Good to know you got it working! Perhaps the touchscreen has increased the noise floor on the audioinjector, just a guess.

**Comment by flatmax** [comment]
In my opinion, because of the reported type of noise, it is frame 
dropping. Frame dropping is most likely due to extra overhead in video 
rendering ...

Doesn't anyone else feel that audio rendering and video rendering should 
be on separate threads ?

Matt
